pynbody/snapshot/pkdgravhdf.py

```diff
@@ -0,0 +1,153 @@
+import configparser
+import logging
+import warnings
+
+from .. import config_parser, family, units
+from .gadgethdf import GadgetHDFSnap, _GadgetHdfMultiFileManager
+
+logger = logging.getLogger('pynbody.snapshot.pkdgravhdf')
+
+try:
+    import h5py
+except ImportError:
+    h5py = None
+
+_pkd_default_type_map = {}
+for x in family.family_names():
+    try:
+        _pkd_default_type_map[family.get_family(x)] = \
+            [q.strip() for q in config_parser.get('pkdgrav3hdf-type-mapping', x).split(",")]
+    except configparser.NoOptionError:
+        pass
+
+_pkd_all_hdf_particle_groups = []
+for hdf_groups in _pkd_default_type_map.values():
+    for hdf_group in hdf_groups:
+        _pkd_all_hdf_particle_groups.append(hdf_group)
+
+
+class _PkdgravHdfMultiFileManager(_GadgetHdfMultiFileManager) :
+    _nfiles_groupname = "Header"
+    _nfiles_attrname = "NumFilesPerSnapshot"
+    _namemapper_config_section = "pkdgrav3hdf-name-mapping"
+
+    def __init__(self, filename, mode='r') :
+        filename = str(filename)
+        self._mode = mode
+        if h5py.is_hdf5(filename):
+            self._filenames = [filename]
+            self._numfiles = 1
+        else:
+            h1 = h5py.File(filename + ".0", mode)
+            self._numfiles = self._get_num_files(h1)
+            if hasattr(self._numfiles, "__len__"):
+                assert len(self._numfiles) == 1
+                self._numfiles = self._numfiles[0]
+            self._filenames = [filename + "."
+                               + str(i) for i in range(self._numfiles)]
+
+        self._open_files = {}
+
+    def get_cosmo_attrs(self):
+        return self[0].parent['Cosmology'].attrs
+
+
+class PkdgravHDFSnap(GadgetHDFSnap):
+    # PKDGRAV3 creates files by default with the format
+    # {name}.{step:05d}.{n}
+    # where `n>=0` is the index of the file when doing parallel writting.
+    # It does not contain any ".hdf5" extension in the file names.
+    _multifile_manager_class = _PkdgravHdfMultiFileManager
+    _readable_hdf5_test_group = "Header"
+    _readable_hdf5_test_attr = "Header", "PKDGRAV version"
+
+    def __init__(self, filename, ):
+        super().__init__(filename)
+
+    def _get_units_from_hdf_attr(self, hdfattrs) :
```

Comment:
If you expect this not to be called, it should raise an exception.

If you do expect it to be called, then it should return something (NoUnit or None perhaps -- unsure what the intent is).